posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,221 @@
+---
+author: 'Yaël Dillies, Paul Lezeau'
+category: 'Metaprogramming'
+date: 2025-04-16 12:00:00 UTC+01:00
+description: 'Informal introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: fantastic-simprocs
+tags: 'simp, simproc, meta'
+title: Fantastic Simprocs and Where to Find Them.
```

Comment:
I know the title is a slight joke, but I would actually be curious to know if there is a way to find or search for simprocs in mathlib? E.g., if I'm in the middle of a proof, is there some way for me to check if there is a simproc that would be helpful here? (Right now it seems to me like I have to keep track of what simprocs mathlib has and what their names are).

Reply:
```lean
run_meta do
  let s ← Lean.Meta.Simp.getSimprocs
  Lean.logInfo m!"{s.pre.toArray ++ s.post.toArray}"
  pure ()
```

Reply:
I think that only finds `simproc`s (the ones enabled automatically), not `simproc_decls` (the ones you refer to by name). Loogle should be able to find both.

Reply:
It would be pretty cool to have a command ``#simprocs` that takes in a pattern and outputs the list of `simproc`s that activate on that pattern, e.g something like `#simprocs (ite _ _ _)` would find `iteReduce`

Reply:
I've changed the title to _Simprocs for the Working (Formal) Mathematician_, which I think make the more practical purpose of this post clearer (and we can repurpose the previous name for the last post in the series where we actually explain how to write a simproc)